sdk/src/main/java/com/hedera/hashgraph/sdk/AddressBookQuery.java

```diff
@@ -0,0 +1,193 @@
+    private FileId fileId = null;
+    @Nullable
+    private Integer limit = null;
+    private int maxAttempts = 10;
+    private Duration maxBackoff = Duration.ofSeconds(8L);
+
+    public AddressBookQuery() {
+    }
+
+    public AddressBookQuery setFileId(FileId fileId) {
+        this.fileId = fileId;
+        return this;
+    }
+
+    @Nullable
+    public FileId getFileId() {
+        return fileId;
+    }
+
+    public AddressBookQuery setLimit(@Nullable @Nonnegative Integer limit) {
+        this.limit = limit;
+        return this;
+    }
+
+    @Nullable
+    public Integer getLimit() {
+        return limit;
+    }
+
+    public AddressBookQuery setMaxAttempts(@Nonnegative int maxAttempts) {
+        this.maxAttempts = maxAttempts;
+        return this;
+    }
+
+    public int getMaxAttempts() {
+        return maxAttempts;
+    }
+
+    public AddressBookQuery setMaxBackoff(Duration maxBackoff) {
+        Objects.requireNonNull(maxBackoff);
+        if (maxBackoff.toMillis() < 500L) {
+            throw new IllegalArgumentException("maxBackoff must be at least 500 ms");
+        }
+        this.maxBackoff = maxBackoff;
+        return this;
+    }
+
+    public NodeAddressBook execute(Client client) {
+        return execute(client, client.getRequestTimeout());
+    }
+
+    public NodeAddressBook execute(Client client, Duration timeout) {
+        var deadline = Deadline.after(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        for (int attempt = 1; true; attempt++) {
+            try {
+                var addressProtoIter = ClientCalls.blockingServerStreamingCall(
+                    buildCall(client, deadline),
+                    buildQuery()
+                );
+                List<NodeAddress> addresses = new ArrayList<>();
+                while (addressProtoIter.hasNext()) {
+                    addresses.add(NodeAddress.fromProtobuf(addressProtoIter.next()));
+                }
+                return new NodeAddressBook().setNodeAddresses(addresses);
+            } catch (Throwable error) {
+                if (!shouldRetry(error) || attempt >= maxAttempts) {
+                    LOGGER.error("Error attempting to get address book at FileId {}", fileId, error);
+                    throw error;
+                }
+                warnAndDelay(attempt, error);
+            }
+        }
+    }
+
+    public CompletableFuture<NodeAddressBook> executeAsync(Client client) {
+        return executeAsync(client, client.getRequestTimeout());
+    }
+
+    public CompletableFuture<NodeAddressBook> executeAsync(Client client, Duration timeout) {
+        var deadline = Deadline.after(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        CompletableFuture<NodeAddressBook> returnFuture = new CompletableFuture<>();
+        executeAsync(client, deadline, returnFuture, 1);
+        return returnFuture;
+    }
+
+    void executeAsync(Client client, Deadline deadline, CompletableFuture<NodeAddressBook> returnFuture, int attempt) {
```

Comment:
There seems a lot of duplication between `TopicMessageQuery` and `AddressBookQuery`. It would be better to use inheritance or composition to share code between the two.

Reply:
I completely agree we should de-duplicate a bunch of this code, but I feel that should be done as a separate task because there are some interesting questions that pop up. I took the liberty of creating a draft PR.